src/Compression/CompressionCodecEncrypted.cpp

```diff
@@ -186,6 +191,138 @@ size_t decrypt(std::string_view ciphertext, char * plaintext, EncryptionMethod m
+/// If something went wrong (can't init context or can't encrypt data) it throws exception.
+/// It returns length of encrypted text.
+size_t encrypt(std::string_view plaintext, char * ciphertext_and_tag, EncryptionMethod method, const String & key, const String & nonce)
+{
+    EVP_CIPHER_CTX *encrypt_ctx;
+
+    if (!(encrypt_ctx = EVP_CIPHER_CTX_new()))
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_cryptinit = EVP_EncryptInit_ex(encrypt_ctx,
+                                                getMethod(method)(),
+                                                nullptr, nullptr, nullptr);
+    if (!ok_cryptinit)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_cipherctrl = EVP_CIPHER_CTX_ctrl(encrypt_ctx,
+                                                  EVP_CTRL_GCM_SET_IVLEN,
+                                                  static_cast<int32_t>(nonce.size()),
+                                                  nullptr);
+    if (!ok_cipherctrl)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_nonceinit = EVP_EncryptInit_ex(encrypt_ctx, nullptr, nullptr,
+                                                reinterpret_cast<const uint8_t*>(key.data()),
+                                                reinterpret_cast<const uint8_t *>(nonce.data()));
+    if (!ok_nonceinit)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    size_t out_len;
+    const int ok_encryptudpate = EVP_EncryptUpdate(encrypt_ctx,
+                                                   reinterpret_cast<uint8_t *>(ciphertext_and_tag),
+                                                   reinterpret_cast<int32_t *>(&out_len),
+                                                   reinterpret_cast<const uint8_t *>(plaintext.data()),
+                                                   static_cast<int32_t>(plaintext.size()));
+    if (!ok_encryptudpate)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_encryptfinal = EVP_EncryptFinal_ex(encrypt_ctx,
+                                                    reinterpret_cast<uint8_t *>(ciphertext_and_tag) + out_len,
+                                                    reinterpret_cast<int32_t *>(&out_len));
+    if (!ok_encryptfinal)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    /* Get the tag */
+    const int ok_tag = EVP_CIPHER_CTX_ctrl(encrypt_ctx,
+                                           EVP_CTRL_GCM_GET_TAG,
+                                           tag_size,
+                                           reinterpret_cast<uint8_t *>(ciphertext_and_tag) + plaintext.size());
+    if (!ok_tag)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    EVP_CIPHER_CTX_free(encrypt_ctx);
+
+    return plaintext.size() + tag_size;
```

Comment:
Are you sure the encrypted size is equal to input text size? shouldn't it be padded to block's multiple plus some bytes for IV? maybe me should use returned size in `out_len`?

Reply:
I wanted to do this initially but in an effort to try to make the encryption as compatible as possible to boringssl, I used the same method for dealing with lengths as boringSSL:
https://github.com/google/boringssl/blob/master/crypto/fipsmodule/cipher/aead.c#L141

Reply:
I don't think it's a good idea - boringssl's writers know internals they are using so they are free to do whatever they consider suitable, however as a library users we are not free to make any such assumptions - there is a paragraphs on openssl doc dedicated to peculiarity of output length:
https://www.openssl.org/docs/man3.0/man3/EVP_EncryptUpdate.html

Reply:
ok, let me see if I can make it work with out length. There is the intricacies of the tag being appended onto the length. I just want to make sure that the SSL providers can decrypt each other encryption.